In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from MyAnalysisApi import setup_env

#setup Enviorment
setup_env()

In [2]:
#load data
df_train = pd.read_csv( "dataset/train/train.csv")
df_test = pd.read_csv(  "dataset/test/test.csv")
df_stores = pd.read_csv(  "dataset/features/stores.csv")
df_sampleSub = pd.read_csv( "dataset/features/sampleSubmission.csv")
df_features = pd.read_csv("dataset/features/features.csv")

In [3]:
print("train dataset\n\n")
df_train.info()
print("test dataset\n\n")
df_test.info()
print("stores dataset\n\n")
df_stores.info()
print("Sample Submission dataset\n\n")
df_sampleSub.info()
print("Features dataset\n\n")
df_features.info()


train dataset


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421570 entries, 0 to 421569
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Store         421570 non-null  int64  
 1   Dept          421570 non-null  int64  
 2   Date          421570 non-null  object 
 3   Weekly_Sales  421570 non-null  float64
 4   IsHoliday     421570 non-null  bool   
dtypes: bool(1), float64(1), int64(2), object(1)
memory usage: 13.3+ MB
test dataset


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115064 entries, 0 to 115063
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   Store      115064 non-null  int64 
 1   Dept       115064 non-null  int64 
 2   Date       115064 non-null  object
 3   IsHoliday  115064 non-null  bool  
dtypes: bool(1), int64(2), object(1)
memory usage: 2.7+ MB
stores dataset


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45 

After look into each dataset, only Features dataset has some missing values, so now we must take a look at each column that has missing values.

In [4]:

df_features.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         8190 non-null   int64  
 1   Date          8190 non-null   object 
 2   Temperature   8190 non-null   float64
 3   Fuel_Price    8190 non-null   float64
 4   MarkDown1     4032 non-null   float64
 5   MarkDown2     2921 non-null   float64
 6   MarkDown3     3613 non-null   float64
 7   MarkDown4     3464 non-null   float64
 8   MarkDown5     4050 non-null   float64
 9   CPI           7605 non-null   float64
 10  Unemployment  7605 non-null   float64
 11  IsHoliday     8190 non-null   bool   
dtypes: bool(1), float64(9), int64(1), object(1)
memory usage: 712.0+ KB


So we can that all markdowns, CPI and Unemployment rate has missing values, but how can we deal with then?


 4   MarkDown1     4032 non-null   float64
 5   MarkDown2     2921 non-null   float64
 6   MarkDown3     3613 non-null   float64
 7   MarkDown4     3464 non-null   float64
 8   MarkDown5     4050 non-null   float64
 9   CPI           7605 non-null   float64
 10  Unemployment  7605 non-null   float64

In [5]:
df_features[["Unemployment", "CPI"]] = df_features[["Unemployment", "CPI"]].interpolate(method="linear")
df_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         8190 non-null   int64  
 1   Date          8190 non-null   object 
 2   Temperature   8190 non-null   float64
 3   Fuel_Price    8190 non-null   float64
 4   MarkDown1     4032 non-null   float64
 5   MarkDown2     2921 non-null   float64
 6   MarkDown3     3613 non-null   float64
 7   MarkDown4     3464 non-null   float64
 8   MarkDown5     4050 non-null   float64
 9   CPI           8190 non-null   float64
 10  Unemployment  8190 non-null   float64
 11  IsHoliday     8190 non-null   bool   
dtypes: bool(1), float64(9), int64(1), object(1)
memory usage: 712.0+ KB


Ok, now we need know how the cler mardowns

In [6]:
miss_vals = [4032,2921,3613,3464]
for i in range(1,4):
    print("Percentage Missing Values Markdown%d:  %.2f%%"%(i, (1 - miss_vals[i]/8190)*100))

Percentage Missing Values Markdown1:  64.33%
Percentage Missing Values Markdown2:  55.89%
Percentage Missing Values Markdown3:  57.70%


As seen above, the percentage of missing values for each markdowwn variable is to high, more than 50% in all cases. If we try to fill or interpolate value we may introduce a big bias on the dataset and compromise the analysis, so I think the best decision is take out these value from analysis.

In [7]:
df_features = df_features.drop(columns=["MarkDown1", "MarkDown2","MarkDown3","MarkDown4","MarkDown5"])


In [8]:
df_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         8190 non-null   int64  
 1   Date          8190 non-null   object 
 2   Temperature   8190 non-null   float64
 3   Fuel_Price    8190 non-null   float64
 4   CPI           8190 non-null   float64
 5   Unemployment  8190 non-null   float64
 6   IsHoliday     8190 non-null   bool   
dtypes: bool(1), float64(4), int64(1), object(1)
memory usage: 392.0+ KB


In [91]:
#saving cleaned data 

df_features.to_csv("dataset/features/features_cleaned.csv")